server/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java

```diff
@@ -193,12 +183,41 @@ public void handleException(final TransportException exp) {
                         });
                     }
                 }
+            } catch (IndexNotFoundException e) {
+                // In some situations it's possible that this is a false exception, i.e. while there's a STATE_NOT_RECOVERED_BLOCK
+                // to ensure that this is a legitimate index not found exception we should check if there's a cluster exception and
+                // handle it if there's one.
+                ClusterBlockException clusterBlockException = checkGlobalBlock(clusterState);
+                if (clusterBlockException != null) {
+                    handleClusterBlockException(clusterState, clusterBlockException);
+                } else {
+                    listener.onFailure(e);
+                }
             } catch (Exception e) {
                 logger.trace("top-level failure", e);
                 listener.onFailure(e);
             }
         }
 
+        private void handleClusterBlockException(ClusterState clusterState, ClusterBlockException blockException) {
+            if (blockException.retryable() == false) {
+                logger.trace("can't execute due to a non-retryable cluster block", blockException);
+                listener.onFailure(blockException);
+            } else {
+                logger.debug("can't execute due to a cluster block, retrying", blockException);
+                retry(clusterState, blockException, newState -> {
+                    try {
+                        ClusterBlockException newException = checkBlock(request, newState);
```

Comment:
If this call to `checkBlock` also throws an INFE and the `STATE_NOT_RECOVERED_BLOCK` is in place then we could reasonably reject the state I think.